### PR TITLE
Add TypeScript definition

### LINKED ISSUE #13: Add TypeScript definition
It'd be great to have a TypeScript definition. Note that there should be a build step to duplicate the definition for the `emittery/legacy` module.

### PR DESCRIPTION
**Pull request by dinoboff**
It including declarations in the npm package using default declaration locations.

Note that by default it doesn’t allow to map an event data type to each event; in TS, Emittery can be aliased to interface to support it. As of TS 2.6, that optional interface doesn’t support symbol.

This PR includes some rather slow tests. To skip those tests, ava can exclude test ending with ‘[slow]’, i.e. `ava —match ‘!*[slow]’`.

Fix #13

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
+// Type definitions for emittery
+// Project: emittery
+// Definitions by: Sindre Sorhus <sindresorhus.com>
+
+export = Emittery;
```
[the remark this answers is not in the document]
Comment (sindresorhus) [reply index.d.ts]: 👍

File: examples/typed.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import Emitter from '../';
```
Comment (novemberborn) [inline examples/typed.ts]: This should import as `Emittery`.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+		/**
+		 * Subscribe to an event.
+		 *
+		 * Using the same listener multiple times for the same event will result
+		 * in only one method call per emitted event.
+		 *
+		 * @returns Unsubscribe method.
+		 */
+		on<U>(eventName: T, listener: (eventData: U) => any): () => void;
```
[the remark this answers is not in the document]
Comment (dinoboff) [reply index.d.ts]: TS will complain if the listener returns / resolves something. Is it worth it?

[the remark this answers is not in the document]
Comment (novemberborn) [reply index.d.ts]: I know, and it's why I'm undecided. I'd like to be more explicit about what happens when a promise is returned, but as you said typing the return value may just get in the way.

**Comment by dinoboff** [comment]
@novemberborn 👍 

I setup typescript in the examples/ folder to not make Typescript a dependency but I don't personally mind.

ps: Is it ok to rebase it here?

**Comment by dinoboff** [comment]
@novemberborn 

Is there anyway to get something like that to work:
```ts
type PickVoid<T> = {
	[P in keyof T]: T[P] & void;
}


class Emittery<Mapped, EventOnly=PickVoid<Mapped>> {
    
    emit<Name extends keyof EventOnly>(eventName: Name, eventData?: undefined): Promise<void>;
    emit<Name extends keyof Mapped>(eventName: Name, eventData: Mapped[Name]): Promise<void> {
        console.log(eventName, eventData);
        return Promise.resolve()
    }
}

const ee = new Emittery<{value: string, close: void}>()

ee.emit('start');        // report error as expected
ee.emit('value');        // Allows it :(
ee.emit('close', 'foo'); // report error as expected
```
[playground](https://www.typescriptlang.org/play/index.html#src=type%20PickVoid%3CT%3E%20%3D%20%7B%0D%0A%09%5BP%20in%20keyof%20T%5D%3A%20T%5BP%5D%20%26%20void%3B%0D%0A%7D%0D%0A%0D%0A%0D%0Aclass%20Emittery%3CMapped%2C%20EventOnly%3DPickVoid%3CMapped%3E%3E%20%7B%0D%0A%20%20%20%20%0D%0A%20%20%20%20emit%3CName%20extends%20keyof%20EventOnly%3E(eventName%3A%20Name%2C%20eventData%3F%3A%20undefined)%3A%20Promise%3Cvoid%3E%3B%0D%0A%20%20%20%20emit%3CName%20extends%20keyof%20Mapped%3E(eventName%3A%20Name%2C%20eventData%3A%20Mapped%5BName%5D)%3A%20Promise%3Cvoid%3E%20%7B%0D%0A%20%20%20%20%20%20%20%20console.log(eventName%2C%20eventData)%3B%0D%0A%20%20%20%20%20%20%20%20return%20Promise.resolve()%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Aconst%20ee%20%3D%20new%20Emittery%3C%7Bvalue%3A%20string%2C%20close%3A%20void%7D%3E()%0D%0A%0D%0Aee.emit('start')%3B%20%20%20%20%20%20%20%20%2F%2F%20report%20error%20as%20expected%0D%0Aee.emit('value')%3B%20%20%20%20%20%20%20%20%2F%2F%20should%20fail%0D%0Aee.emit('close'%2C%20'foo')%3B%20%2F%2F%20report%20error%20as%20expected)

**Comment by novemberborn** [comment]
> ps: Is it ok to rebase it here?

Yea do whatever you need to do. Though we'll probably squash and merge so rebasing is not a necessity.



> Is there anyway to get something like that to work:
> 
> ```
> type PickVoid<T> = {
> 	[P in keyof T]: T[P] & void;
> }
> 
> 
> class Emittery<Mapped, EventOnly=PickVoid<Mapped>> {
>     
>     emit<Name extends keyof EventOnly>(eventName: Name, eventData?: undefined): Promise<void>;
>     emit<Name extends keyof Mapped>(eventName: Name, eventData: Mapped[Name]): Promise<void> {
>         console.log(eventName, eventData);
>         return Promise.resolve()
>     }
> }
> 
> const ee = new Emittery<{value: string, close: void}>()
> 
> ee.emit('start');        // report error as expected
> ee.emit('value');        // Allows it :(
> ee.emit('close', 'foo'); // report error as expected
> ```

This doesn't work because `EventOnly` has the same keys as `Mapped`. `T[P] & void` doesn't matter here.

I think ideally we have `Emittery`, which allows all string event names, with `any` data:

```ts
class Emittery<Mapped = {[eventName: string]: any}, EventOnly = {[eventName: string]: any}> {}
```

Then we'll let you type which events you want to emit, and their data, while still allowing all string event names without data:

```ts
class Example extends Emittery<{foo: 'bar'}> {}
```

And finally we'd let you type which events have no data:

```ts
class Example extends Emittery<{foo: 'bar'}, {baz: true}> {}
```

Unfortunately we can't derive `EventOnly` from `Mapped` by filtering those keys that have `void` values. Still, this seems like the best available solution.

Note that the PR currently uses `EventOnly = Mapped` which disallows unknown event names, but allows all known event names to be used without data. Perhaps that's OK after all?

---

@dinoboff we can't generalize Symbol support like this, though presumably you could add signatures for `on(eventName: symbol, eventData: any)` etc in the subclass. I think that's a decent argument for not supporting symbols at all. What do you think?

**Comment by dinoboff** [comment]
@novemberborn Not necessarily. It can support Symbol even if the type declaration doesn't reflect it; When authoring JS code VSC doesn't complain if I use a symbol as key. Is it an issue in Atom?

An alternative is to define a basic type declaration and provide an interface with an Mapped Event name/data. When authoring in TS could use either depending of the need for symbol.

Event when Symbol support for index lands in TS (TS 2.7 AFAIK), a simpler interface by default might be better for JS authoring.

File: index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dinoboff) [reply index.d.ts]: @novemberborn if defined as an es6 export, VSC's Intellisense in js files expect a default export and only provide autocomplete and description on that default value.

[the remark this answers is not in the document]
Comment (dinoboff) [reply index.d.ts]: ps: i.e. VSC expects `const Emittery = require('emittery').default;`.

[the remark this answers is not in the document]
Comment (novemberborn) [reply index.d.ts]: Ah that makes sense I suppose. Best stick to `export = Emittery` then, since this module assumes CJS not ESM.

**Comment by novemberborn** [comment]
> An alternative is to define a basic type declaration and provide an interface with a Mapped Event name/data. When authoring in TS could use either depending of the need for symbol.
> 
> Even when Symbol support for index lands in TS (TS 2.7 AFAIK), a simpler interface by default might be better for JS authoring.

Yes I like that. Great to hear that TS is improving their symbol support, for now we should have a basic interface that could work for symbols and strings, and as TypeScript improves we can extend the mapped interface to also support symbols.

File: examples/clock.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+		super();
+
+		this._startedAt = 0;
+		this._tick = tick > 0 ? tick : 1000;
+		this._timer = null;
+	}
+
+	async tick() {
+		if (this._tick <= 0) {
+			this.emit('error', new Error('not started'));
```
Comment (novemberborn) [inline examples/clock.js]: Example nitpick, but should this be `await this.emit()`?

File: examples/clock.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	constructor(tick = 1000) {
+		super();
+
+		this._startedAt = 0;
+		this._tick = tick > 0 ? tick : 1000;
+		this._timer = null;
+	}
+
+	async tick() {
+		if (this._tick <= 0) {
```
Comment (novemberborn) [inline examples/clock.js]: Looks like this should check `_startedAt`, or perhaps `_timer === null`?

File: examples/clock.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+			this.emit('error', new Error('not started'));
+			this.stop();
```
Comment (novemberborn) [inline examples/clock.js]: Looks like `this.stop()` will throw in this code path.

File: examples/clock.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+	stop() {
+		if (this._timer === null) {
+			throw new Error('not started');
+		}
+
+		clearInterval(this._timer);
+		this._timer = null;
+		this._startedAt = 0;
+
+		this.emit('started');
```
Comment (novemberborn) [inline examples/clock.js]: `'stopped'`?

File: examples/emit.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env node
+
+'use strict';
+
+const Emittery = require('../');
+
+class MyEmitter extends Emittery {}
```
Comment (novemberborn) [inline examples/emit.js]: Would be nice to use this of an example where you don't need to extend Emittery.

File: index.d.ts [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+		 */
+		listenerCount(eventName?: string): number;
+}
+
+declare namespace Emittery {
+
+	/**
+	 * A map of event names to the data type they emit.
+	 */
+	interface IEvents {
```
Comment (novemberborn) [inline index.d.ts]: Personally I avoid prefixing interfaces with `I`. I suppose the TS community is split on this?

@sindresorhus what do you prefer?

File: index.d.ts [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+		[eventName: string]: any;
+	}
+
+	/**
+	 * Alternative interface for an Emittery object; must list supported events
+	 * the data type they emit.
+	 */
+	export interface IMapped<Events extends IEvents> {
```
Comment (novemberborn) [inline index.d.ts]: I think we should still add the second generic, allowing you to specify which events have no data.

File: test/fixtures/compiles/off.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import Emittery = require('../../..');
+
+const ee = new Emittery();
+const listener = () => undefined;
```
Comment (novemberborn) [inline test/fixtures/compiles/off.ts]: `listener` is unused.

File: test/types.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+	t.is(errors.length, 0, errorMessage(errors));
+}
+
+function assertEachFail(t, srcDir) {
+	const fileNames = listFiles(resolve(srcDir)).sort();
+	const errors = compile(fileNames);
+	const filesWithErrors = errors
+		.map(err => (err.file ? err.file.fileName : null))
+		.filter(Boolean)
+		.filter(unique);
```
Comment (novemberborn) [inline test/types.js]: You could pass the boolean-filtered array to `new Set()` to get unique values. If you do the same with `fileNames` you can still use `t.deepEqual()`.

File: index.d.ts (hunk 9 above)
Comment (dinoboff) [reply index.d.ts]: I don't mind. Just following TSlint advice.

Comment (sindresorhus) [reply index.d.ts]: I prefer without the prefix too.

**Comment by novemberborn** [comment]
@dinoboff I've done some iterations on your PR, please see the commits I've just pushed. I think I've got a nicer approach for mapping event types now.

I'm not sure what to do with the documentation — VS Code and Atom show much less detail than I'd hoped, especially for `Emittery.Typed`. We'd need to duplicate the `README` twice more to have docs for all methods in all use cases. Is it even worth having these docs in the TypeScript definition?

@sindresorhus this adds an empty subclass to `Emittery` itself. Please let me know what you think about that. It's the best way I've found of supporting generic `Emittery` usage while encouraging people to type their events. I also reworded `off()` in the docs.

**Comment by sindresorhus** [comment]
> I'm not sure what to do with the documentation — VS Code and Atom show much less detail than I'd hoped, especially for Emittery.Typed.

What is it missing? Maybe we could open some issues about it on those projects or the TS issue tracker.

And can you quickly mention `Emittery.Typed` in the readme for TS users? So they know upfront what they should use. You can just refer to the type docs for more info.

> this adds an empty subclass to Emittery itself. Please let me know what you think about that. It's the best way I've found of supporting generic Emittery usage while encouraging people to type their events. I also reworded off() in the docs.

Yeah, I'm ok with that.

**Comment by novemberborn** [comment]
>> I'm not sure what to do with the documentation — VS Code and Atom show much less detail than I'd hoped, especially for Emittery.Typed.
> 
> What is it missing? Maybe we could open some issues about it on those projects or the TS issue tracker.

The `Emittery` import doesn't show any documentation, though `Emittery.Typed`'s class documentation does show up. But none of the method overloads of `Emittery.Typed` have documentation. I suppose that's somewhat sensible being an overload and all but if we'd want that to be visible we'd have to duplicate the docs for each overload.

> And can you quickly mention Emittery.Typed in the readme for TS users? So they know upfront what they should use. You can just refer to the type docs for more info

Yes good catch.

---

I'm having second thoughts on how `legacy.d.ts` simply aliases `Emittery.d.ts`. It leads to this problem:

```ts
import E = require('.')
import L = require('./legacy')

let v: E
v = new L()
console.log(v instanceof E)
```

This passes type checks, but of course the classes are different so `v instanceof E` is actually `false`.  Functionally both classes are the same though. Perhaps we can live with this edge case? @dinoboff what do you think?

**Comment by dinoboff** [comment]
@novemberborn I am not sure about the issue. I am guessing you would prefer TS to complain about `typeof`?

**Comment by novemberborn** [comment]
> @novemberborn I am not sure about the issue. I am guessing you would prefer TS to complain about typeof?

@dinoboff yea. But turns out even if I copy `Emittery.d.ts` into `legacy.d.ts` my code snippet above behaves the same way, so I don't think this is actually an issue.

**Comment by novemberborn** [comment]
@dinoboff @sindresorhus I think this is ready now. I've left the method documentation in the definition file for now, let me know if you'd want me to remove this instead.

**Comment by sindresorhus** [comment]
@novemberborn Can you land it in two commits, one commit with @dinoboff's commits squashed and one with yours?

**Comment by novemberborn** [comment]
Thanks for kickstarting this, as well as being so appreciative of the back and forth iterations @dinoboff!

**Comment by sindresorhus** [comment]
Thanks @dinoboff and @novemberborn. This is amazing :)

**Comment by dinoboff** [comment]
Likewise @novemberborn, thanks.

**Comment by sindresorhus** [comment]
@dinoboff Would you be interested in joining the project as a maintainer? @novemberborn and I would be happy to have you if you're interested :)

**Pull request closed without merging**